FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/tools/tcp_dump_packets.c

```diff
@@ -0,0 +1,657 @@
+#define dnsDNS_PORT						0x0035u
+
+/* Some const values describing the 'flags' in a TCP packet. */
+#define tcpTCP_FLAG_FIN					0x0001u /* No more data from sender */
+#define tcpTCP_FLAG_SYN					0x0002u /* Synchronize sequence numbers */
+#define tcpTCP_FLAG_RST					0x0004u /* Reset the connection */
+#define tcpTCP_FLAG_PSH					0x0008u /* Push function: please push buffered data to the recv application */
+#define tcpTCP_FLAG_ACK					0x0010u /* Acknowledgment field is significant */
+
+/* A macro to add a type, both as a numeric value, as well as a string. */
+#define ADD_TYPE( FLAGS ) \
+	prvAddType( flag_##FLAGS, #FLAGS )
+
+/*-----------------------------------------------------------*/
+
+static char pcTypeString[ 255 ];
+static uint32_t ulTypeMask;
+
+/* The name of the C source file to be written. */
+static char pcCodeFileName[ MAX_PATH ];
+
+/* The name of the header file to be written. */
+static char pcHeaderFileName[ MAX_PATH ];
+
+/* A stream buffer between the FreeRTOS tasks and the Windows thread. */
+static StreamBuffer_t *xPacketBuffer;
+
+/* A process handle of the Windows thread. */
+static HANDLE pvProcessHandle;
+
+static UBaseType_t uxNextPacketNumber;
+static BaseType_t xFirstPacket = 1;
+
+/* Bollean 'xDumpingReady' becomes true once all desired packet have been collected.
+Further packets will be dropped (ignored). */
+static volatile BaseType_t xDumpingReady = pdFALSE;
+
+static DumpEntries_t *pxCurrentEntries;
+
+static uint16_t usSourcePort;
+static uint16_t usDestinationPort;
+
+typedef struct xBufferheader
+{
+	size_t uxLength;
+	UBaseType_t bIncoming : 1;
+} Bufferheader_t;
+
+static DumpEntries_t xExampleEntries = {
+	.uxEntryCount = 4,	/* No more than 'dumpMAX_DUMP_ENTRIES' elements. */
+	.xEntries = {
+		{ .ulMask = flag_IN | flag_UDP,   .uxMax = 2u },
+		{ .ulMask = flag_IN | flag_ARP,   .uxMax = 2u },
+		{ .ulMask = flag_IN | flag_TCP,   .uxMax = 5u },
+		{ .ulMask = flag_IN | flag_SYN,   .uxMax = 1u },
+	}
+};
+
+const char pcHeaderCode[] =
+	"/*\n"
+	" * This file was created automatically by 'dump_packets.c'\n"
+	" */\n"
+	"\n"
+	"/* Standard includes. */\n"
+	"#include <stdio.h>\n"
+	"#include <stdint.h>\n"
+	"#include <stdarg.h>\n"
+	"#include <io.h>\n"
+	"#include <ctype.h>\n"
+	"\n"
+	"/* FreeRTOS includes. */\n"
+	"#include <FreeRTOS.h>\n"
+	"#include <task.h>\n\n"
+	"#include \"%s\"\n\n";
+
+const char pcHeaderHeader[] =
+	"/*\n"
+	" * This file was created automatically by 'dump_packets.c'\n"
+	" */\n"
+	"\n"
+	"#ifndef PACKET_LIST_H\n\n"
+	"#define PACKET_LIST_H\n\n"
+	"typedef struct xDumpPacket\n"
+	"{\n"
+	"	const uint8_t *pucData;\n"
+	"	size_t uxLength;\n"
+	"	uint32_t ulType;\n"
+	"	uint16_t usSource;\n"
+	"	uint16_t usDestination;\n"
+	"} DumpPacket_t;\n\n";
+
+/*-----------------------------------------------------------*/
+
+/* The Windows thread that actually writes the network packets to a C source and header file. */
+static DWORD WINAPI prvWritePackets( LPVOID lpParameter );
```

Comment:
Could we use some abstraction so this code can be reused on other platforms? I am going to approve, but this could be useful for the Linux port

Reply:
I suppose we may as well shelf that for later, as the Linux port does not have networking added to it yet.

Reply:
We can add that later, as soon as the Linux port also has TCP.

And there is also FreeRTOS+FAT: the module `tcp_dump_packets.c` could also run on any device that has a disk ( SD/MMC, or a RAM-disk ).
In that case, mentioned function would become a normal task:
~~~c
    static void prvWritePackets( void *pvParameter );
~~~
I have thought of adding some +FAT code, but the problem is that there are no AWS projects yet that use FreeRTOS+FAT.
( and once you have a disk, it would also be nice to be able to produce regular PCAP files ).

The `tcp_mem_stats.c` module, on the contrary, can run on any device that has logging.